apollo-federation/tests/extract_subgraphs.rs

```diff
@@ -255,3 +255,233 @@ fn erase_empty_types_due_to_overridden_fields() {
+      
+      directive @join__field(graph: join__Graph, requires: join__FieldSet, provides: join__FieldSet, type: String, external: Boolean, override: String, usedOverridden: Boolean, overrideLabel: String, contextArguments: [join__ContextArgument!]) repeatable on FIELD_DEFINITION | INPUT_FIELD_DEFINITION
+      
+      directive @join__graph(name: String!, url: String!) on ENUM_VALUE
+      
+      directive @join__implements(graph: join__Graph!, interface: String!) repeatable on OBJECT | INTERFACE
+      
+      directive @join__type(graph: join__Graph!, key: join__FieldSet, extension: Boolean! = false, resolvable: Boolean! = true, isInterfaceObject: Boolean! = false) repeatable on OBJECT | INTERFACE | UNION | ENUM | INPUT_OBJECT | SCALAR
+      
+      directive @join__unionMember(graph: join__Graph!, member: String!) repeatable on UNION
+      
+      directive @link(url: String, as: String, for: link__Purpose, import: [link__Import]) repeatable on SCHEMA
+      
+      directive @listSize(assumedSize: Int, slicingArguments: [String!], sizedFields: [String!], requireOneSlicingArgument: Boolean = true) on FIELD_DEFINITION
+      
+      enum AorB
+        @join__type(graph: SUBGRAPHWITHCOST)
+        @cost(weight: 15)
+      {
+        A @join__enumValue(graph: SUBGRAPHWITHCOST)
+        B @join__enumValue(graph: SUBGRAPHWITHCOST)
+      }
+      
+      type HasInts
+        @join__type(graph: SUBGRAPHWITHLISTSIZE)
+      {
+        ints: [Int!]
+      }
+      
+      input InputTypeWithCost
+        @join__type(graph: SUBGRAPHWITHCOST)
+      {
+        somethingWithCost: Int @cost(weight: 20)
+      }
+      
+      input join__ContextArgument {
+        name: String!
+        type: String!
+        context: String!
+        selection: join__FieldValue!
+      }
+      
+      scalar join__DirectiveArguments
+      
+      scalar join__FieldSet
+      
+      scalar join__FieldValue
+      
+      enum join__Graph {
+        SUBGRAPHWITHCOST @join__graph(name: "subgraphWithCost", url: "")
+        SUBGRAPHWITHLISTSIZE @join__graph(name: "subgraphWithListSize", url: "")
+      }
+      
+      scalar link__Import
+      
+      enum link__Purpose {
+        """
+        `SECURITY` features provide metadata necessary to securely resolve fields.
+        """
+        SECURITY
+      
+        """
+        `EXECUTION` features provide metadata necessary for operation execution.
+        """
+        EXECUTION
+      }
+      
+      type Query
+        @join__type(graph: SUBGRAPHWITHCOST)
+        @join__type(graph: SUBGRAPHWITHLISTSIZE)
+      {
+        fieldWithCost: Int @join__field(graph: SUBGRAPHWITHCOST) @cost(weight: 5)
+        argWithCost(arg: Int @cost(weight: 10)): Int @join__field(graph: SUBGRAPHWITHCOST)
+        enumWithCost: AorB @join__field(graph: SUBGRAPHWITHCOST)
+        inputWithCost(someInput: InputTypeWithCost): Int @join__field(graph: SUBGRAPHWITHCOST)
```

Comment:
can we also add Object and Scalar types with `@cost` definition to this test?

Reply:
Uh oh. My lack of GraphQL knowledge is showing 😅 

Noob question: how does Object/Scalar differ from a field that is/returns an object type or scalar value?

Reply:
In the case of the `@cost` directive definition that’s written out above, the DirectiveLocation refers to the Object Type Definition, which is [this](https://spec.graphql.org/October2021/#sec-Objects). So any directive actually defined in the `Directives` part of that grammar. So in practice what the cost definition on an Object would mean is that any time you’d have a field definition returning that Object, you wouldn’t have to specify it for every field, you’d just look up the cost on the referenced Object type.

Same principles would apply to [Scalar Type Defintions](https://spec.graphql.org/October2021/#sec-Scalars). 

Does that make sense?

Reply:
Updated!